### ISSUE TITLE
`SoftmaxOverSpatialLayer` might produce NaN gradients

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#576]
> > > Note that it's really important to never mask away all keys for a single query, even if the query itself is inside padding and might be masked away later.
> > > This will cause the softmax to return `NaN` for that query pos, causing the loss for that query also to be `NaN`. And even though it might be masked away when the total loss is accumulated, this can still cause the total loss to be `NaN`. I had this issue sometimes. That's not difficult to fix (e.g. `SoftmaxOverSpatialLayer` could never apply a softmax for masked away query positions), but maybe we should keep it in mind.
> > 
> > 
> > I think this is somewhat orthogonal to this issue, so we should open a separate issue and discuss over there.
> > As I understand you, this happened for you for frames which are inside the padding (i.e. outside of the sequence).
> > I'm not exactly sure how it can happen then that this ends up in the final loss, because for framewise losses (e.g. CE), it would cut out only the sequences, i.e. remove the padded frames, and then sum only over the real relevant frames.
> 
> The problem is that this "cutting to only the sequences" is implemented with `tf.where`, and that the gradient of `tf.where` is NaN as soon as one input is NaN, even if it is masked away. From the [tf.where](https://www.tensorflow.org/api_docs/python/tf/where) docs:
> 
> > Note that if the gradient of either branch of the tf.where generates a NaN, then the gradient of the entire tf.where will be NaN

We even have `test_where_nan` about that.

> > But anyway, let's discuss this in a separate issue.
> 
> I can make it an issue, but it's more a TF issue than a RETURNN issue (actually see [tensorflow/tensorflow#38349](https://github.com/tensorflow/tensorflow/issues/38349)). I just mentioned this because once/if we add some implicit "HistoryMaskLayer", then this is something we also should implicitly handle. Right now, you can just handle this all explicitly.

No, it is a RETURNN issue. There should be no NaN in the loss or gradient when it is incorrect. Then it just means we need to add some workaround. But anyway, this is independent of this issue here, so in any case, it should be discussed elsewhere, to not blow up this issue thread even more.

_Originally posted by @albertz / @Zettelkasten in https://github.com/rwth-i6/returnn/issues/391#issuecomment-896882118_
